Add missing cbor round trip tests

teodanciu opened the pull request:
# Description
In this PR I'm adding: 
*  roundtrip tests for:  TxDats, Remeemers 
*  cbor roundtrip tests for Block for eras [Shelley .. Alonzo]
*  cddl roundtrip tests for BHeader and BHBody for eras [Allegra  ..  Alonzo]
*  huddle roundtrip tests BHeader and BHBody for eras [Allegra. .. Alonzo]
* functions that test cddl and huddle roundtrip tests for Babbage era and beyond, to be called by Consensus

Also, removing a duplicated roundtrip check.

Also, i have discovered that the cddl for babbage for bheader is incorrent: the operational certificate is serialized like in conway, and not like in Shelley-Alonzo, so  I have fixed the cddl for babbage. 

Below is an example of  why what we have in master is wrong: 

1. A cbor representation of BHeader in [Shelley..Alonzo]: 
```
[[
	[8640, 172836,  ; block_number, slot
	h'..',    		; block_number, slot
	h'..',    		; issuer_vkey
	h'..',    		; vrf_vkey
	[h'..', h'..'], ; nonce_vrf
	[h'..', h'..'], ; leader_vrf
	4,  			; block_body_size
	h'..',			; block_body_hash
	h'..', 0, 0, h'..', 7, 2  ;operational certificate & prot version
]]
```

2. A cbor representation of BHeader starting with Babbage: 
```
[[
	89439, 1912624, ; block_number, slot
	h'..', 			; block_number, slot
	h'..',			; issuer_vkey
	h'..', 			; vrf_vkey
	[h'..', h'..'], ; vrf_result
	4,				; block_body_size
	h'..', 			; block_body_hash
	[h'..', 0, 13, h'..'], ; operational_cert
	[8, 0]			
]]
```

The change I made to babbage cddl is aligning the definition to the actual deserialization. 

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commits in meaningful sequence and with useful messages
- [ ] Tests added or updated when needed
- [ ] `CHANGELOG.md` files updated for packages with externally visible changes<br>
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] Version bounds in `.cabal` files updated when necessary<br>
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code formatted (use `scripts/fourmolize.sh`)
- [x] Cabal files formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Add cbor round trip tests for `Block` for each era

lehins commented on libs/cardano-protocol-tpraos/test/Test/Cardano/Protocol/Binary/RoundTrip.hs:
```diff
@@ -0,0 +1,46 @@
+roundTripBlock =
+  prop (show (typeRep $ Proxy @(Block (BHeader StandardCrypto) era))) $
+    withMaxSuccess 3 $
+      roundTripAnnEraExpectation @era @(Block (BHeader StandardCrypto) era)
```
It is better to add this funciton to the testlib, so it can be used in consensus as well. For this it needs to be parameterized by block header
```suggestion
roundTripBlock ::
  forall era bh.
  ( EraSegWits era
  , Arbitrary (Tx era)
  ) =>
  Spec
roundTripBlock =
  prop (show (typeRep $ Proxy @(Block bh era))) $
    withMaxSuccess 3 $
      roundTripAnnEraExpectation @era @(Block bh era)
```

teodanciu replied:
Sure! good point, thanks.

lehins commented on libs/cardano-protocol-tpraos/test/Test/Cardano/Protocol/Binary/RoundTrip.hs:
```diff
@@ -0,0 +1,46 @@
+    roundTripBlock @MaryEra
+    roundTripBlock @AlonzoEra
+    roundTripBlock @BabbageEra
+    roundTripBlock @ConwayEra
```
These do not make sense, because block header as defined in this package is never used with Babbage onwards
```suggestion
```

teodanciu replied:
Yeah I wondered about whether it brings any value at all like this, but  I agree, no point testing unrealistic stuff. 
I will try adding for TxSeq instead!

teodanciu committed:
Add missing roundtrip tests for `TxDats` and `Redeemers`

in alonzo and babbage.

teodanciu commented on libs/cardano-protocol-tpraos/test/Test/Cardano/Protocol/Binary/CddlSpec.hs:
```diff
@@ -27,20 +28,25 @@ spec :: Spec
+      huddleRoundTripAnnCborSpec @(BHeader StandardCrypto) v "header"
+      huddleRoundTripCborSpec @(BHeader StandardCrypto) v "header"
+      huddleRoundTripCborSpec @(BHBody StandardCrypto) v "header_body"
+      xdescribe "[ operational_cert ] doesn't roundtrip with huddle" $
```
Is this a known issue? It does ring a bell, but couldn't find a github issue. If there isn't one, I will create one.

lehins replied:
There was a relevant ticket, but I thought it was fixed: https://github.com/IntersectMBO/cardano-ledger/issues/3559

Feel free to create a new one and figure it out what is going on with it.

teodanciu committed:
Fix cddl for `operational_cert` for babbage

so that it conforms to the actual deserialization

teodanciu committed:
Add missing cddl roundtrip tests for post-shelley eras

as well as defining post-babbage test functions to be called by
consensus, since the header-types are defined there rather than in
ledger

teodanciu committed:
Add huddle roundtrip tests for post-shelley eras

and define functions to be called for post-babbage eras from consensus

teodanciu committed:
Remove duplicated roundtrip tests for `BootstrapWitness`

since they are already called in core `BinarySpec`

teodanciu commented on libs/cardano-protocol-tpraos/test/Test/Cardano/Protocol/Binary/CddlSpec.hs:
```diff
@@ -1,46 +1,68 @@
+    describe "Huddle" $
+      specWithHuddle cddlFiles 100 $ do
+        huddleEraSpec @era @StandardCrypto @BHeader @BHBody
+        xdescribe "Cannot generate a CBOR term corresponding to a group with cuddle" $
```
@nc6  -  Could you confirm if this is by design like this, or if it's just something pending atm in cuddle?

lehins commented on libs/cardano-protocol-tpraos/testlib/Test/Cardano/Protocol/Binary/Cddl.hs:
```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```
It is usually a bad idea to disable this warning. Is there a reason why we can't delete redundant constraints?
```suggestion
```

teodanciu replied:
Same, because of  this constraint :  
`  , AtLeastEra BabbageEra era` in  `postBabbageCddlSpec` and `postBabbageHuddleSpec` signatures. 
If there is no trick to make it not complain, I will remove them.

lehins commented on libs/cardano-protocol-tpraos/testlib/Test/Cardano/Protocol/Binary/Cddl.hs:
```diff
@@ -0,0 +1,113 @@
+  cddlEraSpec,
+  huddleEraSpec,
+  postBabbageCddlSpec,
+  postBabbageHuddleSpec,
```
Doesn't "post"Babbage implies that it is applicable only after Babbage?
We could use `babbageOnwards`, but I think a better use the terminology that consensus uses?

```suggestion
  praosBlockCddlSpec,
  praosBlockHuddleSpec,
```

teodanciu replied:
I went back and forth about it. In the end i found some places in our code where "postBabbage" includes babbage, but I agree, it's ambiguous -  and it's ambiguous in the worst way possible - basically unclear about exactly what it wants to be precise about!!  
Thank you for the suggestion, I like it! :+1:

lehins replied:
> i found some places in our code where "postBabbage" includes babbage,

I would suggest avoiding using cardano-ledger-test as an example of good naming conventions :smiling_face_with_tear:

lehins commented on libs/cardano-protocol-tpraos/testlib/Test/Cardano/Protocol/Binary/Cddl.hs:
```diff
@@ -0,0 +1,113 @@
+  , EncCBOR (bhbody c)
+  ) =>
+  SpecWith CddlData
+cddlEraSpec = do
```
I think adding a Block suffix will make it a bit more clear which spec we have in mind here
```suggestion
cddlEraBlockSpec = do
```

lehins commented on libs/cardano-protocol-tpraos/testlib/Test/Cardano/Protocol/Binary/RoundTrip.hs:
```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
```
There aren't any instances here
```suggestion
```

teodanciu replied:
Indeed! I must have copied all these from another file and forgot to remove them. Thank you

lehins commented on libs/cardano-protocol-tpraos/testlib/Test/Cardano/Protocol/Binary/RoundTrip.hs:
```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
```
I might be wrong, but it doesn't look like type families are used here
```suggestion
```

lehins commented on libs/cardano-protocol-tpraos/test/Test/Cardano/Protocol/Binary/CddlSpec.hs:
```diff
@@ -1,46 +1,68 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```
Same here. Is there a really good reason why we need to disable this warning?
```suggestion
```

teodanciu replied:
I added it because of the `AtMostEra` constraint, which I wanted to add to the signature in order to clarify that the tests are not valid for eras after Alonzo:
```specForEra ::  forall era. (Era era, AtMostEra AlonzoEra era) =>```

But I agree, disabling the warning might cause bigger problems than this solves.
I wonder how it works without warnings in other cases, for example in  [TxInfo](https://github.com/IntersectMBO/cardano-ledger/blob/master/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/TxInfo.hs#L331):
```
transTxCert :: (ShelleyEraTxCert era, ProtVerAtMost era 8) => TxCert era -> PV1.DCert
transTxCert txCert 
```

Is the usage within the function somehow implying at most  8 Protocol version? 

Anyway, I'm happy to remove the constraint and add a haddock or change the function name if there is no way to use this constraint without disabling the warning.

lehins replied:
I see. Normally this constraint will not be complained about because some function is being used that requires this constraint. For example in case of `transTxCert` it is the case that one of the deprecated constructors is being used that requires that constraint. In this case there is the trick you can use to have your cake and it too :smile: There is no need to using this hammer of disabling the warning
https://github.com/IntersectMBO/cardano-ledger/blob/92d93abee27ee1ddd779ce6e63e54e0c44664bc4/libs/cardano-ledger-api/test/Test/Cardano/Ledger/Api/Tx/Out.hs#L42

teodanciu replied:
Great! Replaced hammer with cunning trick, thank you!

lehins reviewed (approved):
Very nice piece of work. Thank you for finding this bug in Babbage cddl
